eth/watchers/senderwatcher_test.go

```diff
@@ -0,0 +1,571 @@
+	}
+	watcher := &stubBlockWatcher{}
+	sw, err := NewSenderWatcher(stubTicketBrokerAddr, watcher, lpEth)
+	assert.Nil(err)
+
+	header := defaultMiniHeader()
+	newWithdrawalEvent := newStubWithdrawalLog()
+	header.Logs = append(header.Logs, newWithdrawalEvent)
+
+	blockEvent := &blockwatch.Event{
+		Type:        blockwatch.Added,
+		BlockHeader: header,
+	}
+
+	go sw.Watch()
+	defer sw.Stop()
+	time.Sleep(2 * time.Millisecond)
+
+	// If map entry is nil don't handle event
+	watcher.sink <- []*blockwatch.Event{blockEvent}
+	time.Sleep(2 * time.Millisecond)
+	_, ok := sw.senders[stubSender]
+	assert.False(ok)
+
+	// If map entry exists parse event log
+	info, err := sw.GetSenderInfo(stubSender)
+	require.Nil(err)
+	require.Equal(info, senderInfo)
+	time.Sleep(2 * time.Millisecond)
+	watcher.sink <- []*blockwatch.Event{blockEvent}
+	time.Sleep(2 * time.Millisecond)
+	info, err = sw.GetSenderInfo(stubSender)
+	assert.Nil(err)
+	assert.Zero(info.Deposit.Cmp(big.NewInt(0)))
+	assert.Zero(info.Reserve.Cmp(big.NewInt(0)))
+
+	// if block removed, make RPC call
+	blockEvent.Type = blockwatch.Removed
+	info, err = sw.GetSenderInfo(stubSender)
+	assert.Nil(err)
+	assert.Zero(info.Deposit.Cmp(big.NewInt(0)))
+	assert.Zero(info.Reserve.Cmp(big.NewInt(0)))
+
+	// If we don't care about the address, don't handle the event
+	s := pm.RandAddress()
+	sender := ethcommon.LeftPadBytes(s.Bytes(), 32)
+	var senderTopic ethcommon.Hash
+	copy(senderTopic[:], sender[:])
+	newWithdrawalEvent.Topics[1] = senderTopic
+	watcher.sink <- []*blockwatch.Event{blockEvent}
+	time.Sleep(2 * time.Millisecond)
+	_, ok = sw.senders[s]
+	assert.False(ok)
+}
+
+func TestWinningTicketTransferEvent(t *testing.T) {
+	assert := assert.New(t)
+	require := require.New(t)
+	deposit, _ := new(big.Int).SetString("10000000000000000000", 10)
+	reserve, _ := new(big.Int).SetString("5000000000000000000", 10)
+
+	senderInfo := &pm.SenderInfo{
+		Deposit: deposit,
+		Reserve: reserve,
+	}
+	lpEth := &eth.StubClient{
+		SenderInfo: senderInfo,
+	}
+	watcher := &stubBlockWatcher{}
+	sw, err := NewSenderWatcher(stubTicketBrokerAddr, watcher, lpEth)
+	require.Nil(err)
+
+	header := defaultMiniHeader()
+	newWinningTicketEvent := newStubWinningTicketLog()
+	header.Logs = append(header.Logs, newWinningTicketEvent)
+
+	blockEvent := &blockwatch.Event{
+		Type:        blockwatch.Added,
+		BlockHeader: header,
+	}
+
+	go sw.Watch()
+	defer sw.Stop()
+	time.Sleep(2 * time.Millisecond)
+
+	// If map entry is nil don't handle event
+	watcher.sink <- []*blockwatch.Event{blockEvent}
+	time.Sleep(2 * time.Millisecond)
+	info, ok := sw.senders[stubSender]
+	assert.False(ok)
+
+	// If map entry exists parse event log
+	// ticket amount is 200 gwei
+	// init the map entry
+	info, err = sw.GetSenderInfo(stubSender)
+	require.Nil(err)
+	require.Equal(info, senderInfo)
+	assert.Zero(info.Deposit.Cmp(senderInfo.Deposit))
+	assert.Zero(info.Reserve.Cmp(senderInfo.Reserve))
+	faceValue := big.NewInt(200000000000)
```

Comment:
We should also have a test case for when the ticket face value exceeds the deposit so that the redemption drains the deposit to 0 and subtracts the remaining face value from the reserve

Reply:
fixed in 6e29515